### PR TITLE
Fix groups with no tasks

### PR DESCRIPTION
**Pull request by Hagellach37**
This PR fixes the grouping workflow for TMS projects ("normal" build area projects in the app)

* make sure that there are no groups without tasks #434 
* make sure that the group_x_size is always even, to make sure that a screen always shows 6 tiles #271 
* remove all overlapping groups which still occurred when using Multipolygons for project creation #374

### CODE REVIEW

File: mapswipe_workers/mapswipe_workers/utils/tile_grouping_functions.py [hunk 1]
```diff
@@ -291,6 +299,53 @@ def groups_intersect():
             and (int(y_minB) <= int(y_max))
         )
 
+    def merge_groups():
```
Comment (laurentS) [inline mapswipe_workers/mapswipe_workers/utils/tile_grouping_functions.py:302]: I would suggest making this a pure function, that takes 2 groups (or whatever you need) as arguments and returns a new group, and does not use any external variables. Right now, you're relying on context to give you the input data, and this tends to be very fragile, and therefore bug prone. If you move this function elsewhere in your code, or you change your code before or after this, you might change the meaning of a variable, and have to update the body of this function. You're breaking the "encapsulation" of this function in doing so, and it often leads to sad developers :)

File: mapswipe_workers/mapswipe_workers/utils/tile_grouping_functions.py [hunk 2]
```diff
@@ -318,55 +373,17 @@ def groups_intersect():
 
             if groups_intersect():
```
Comment (laurentS) [inline mapswipe_workers/mapswipe_workers/utils/tile_grouping_functions.py:374]: Same comment as above: I would make this a pure function (pure meaning it has no side effect and all the data it needs is clearly passed through arguments).

**Review by laurentS — APPROVED** [review]
Looks ok to me, aside from a comment about making functions better encapsulated. This is mostly to limit the risk of bugs creeping in later on.
I'm also wondering why you create groups and then filter them to remove odd-length ones. Is there no way to make sure that their size is even when initially creating them? It seems like it would be less expensive computationally.

**Comment by Hagellach37** [comment]
Hey @laurentS,
one quick comment on why there is a need to filter the odd length groups. In the first place we create groups that have an even length. But: if groups overlap by a single tile, we just merge these two groups into a single group. But this merge operation can result in an odd group length. That's why we have to check again after merging groups.

I will address you other comments. Thanks for the suggestions. :)

**Pull request merged**